core/gather/driver/target-manager.js

```diff
@@ -168,6 +170,9 @@ class TargetManager extends ProtocolEventEmitter {
       // Cast because tsc 4.7 still can't quite track the dependent parameters.
       const payload = /** @type {LH.Protocol.RawEventMessage} */ ({method, params, sessionId});
       this.emit('protocolevent', payload);
+      if (method === 'Target.attachedToTarget') {
+        this._targetAttachedPayloads.push(payload);
+      }
```

Comment:
I _was_ thinking instead of having to hold onto them, we could recreate the target events from `this._targetIdToTargets.values()`, since all that info is in there, since that means if/when we decide to more actively attach/detach from targets (we don't currently do anything for `Target.targetCrashed`/`Target.targetDestroyed`/`Target.targetInfoChanged`/`Target.detachedFromTarget` events), this approach would automatically pick up any management work and only emit known targets.

However, that would mean any `Target` events that come in during the run wouldn't get the same treatment, so to be consistent we'd either need to save all `Target.*` events here and downstream consumers like `network-recorder` would have to be able to handle evolving targets, or `TargetManager` could handle the whole thing, but it probably isn't a good idea to rely on events at that point.

Alternatively we change `_getProtocolEventListener` above to take a `targetType`, so payload becomes `{method, params, targetType}` (or also keep `sessionId`, but there really wouldn't be any need for it at that point), `_findRealRequestAndSetSession` becomes `_findRealRequestAndSetTargetType`, and no other special handling is needed.

Reply:
It sounds like re-emitting is simpler, and we could defer changing the interface for target manager / having downstream consumers do more work until we actually need it?

> Alternatively we change _getProtocolEventListener above to take a targetType, so payload becomes {method, params, targetType} (or also keep sessionId, but there really wouldn't be any need for it at that point), _findRealRequestAndSetSession becomes _findRealRequestAndSetTargetType, and no other special handling is needed.

This loses the property that the network records can be created from just the devtools log. Or are you suggesting we also save targetType in DevToolsLog / what we write to disk?

Reply:
> Or are you suggesting we also save targetType in DevToolsLog / what we write to disk?

yes, if it's in `payload` here it'll be automatically saved in the devtoolslog (like sessionId is currently)

Reply:
From chat...

I like the alternative plan, assuming we store DT logs as `{method, params, targetType, sessionId}` instead of the current `{method, params, targetType}`. I would prefer maintaining our new definition of protocol events as an extended version of the original.